[LOOP-1315] Upload Alert history to Tidepool backend

darinkrauss opened the pull request:
- https://tidepool.atlassian.net/browse/LOOP-1315
- Refactor AlertStore query and filter to not return managed objects
- Add AlertStoreDelegate to notify delegate when alerts are updated
- Add parameters property to Alert and StoredAlert
- Add syncIdentifier property to StoredAlert
- Update tests

Commit:
[LOOP-1315] Upload Alert history to Tidepool backend

- https://tidepool.atlassian.net/browse/LOOP-1315
- Refactor AlertStore query and filter to not return managed objects
- Add AlertStoreDelegate to notify delegate when alerts are updated
- Add parameters property to Alert and StoredAlert
- Add syncIdentifier property to StoredAlert
- Update tests

darinkrauss commented:
LW PR for build: https://github.com/tidepool-org/LoopWorkspace/pull/757

rickpasetto commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -343,46 +350,59 @@ extension AlertStore {
+            return
+        }
+
+        self.managedObjectContext.performAndWait {
```
Just wondering, why did you switch to using `perform` instead of `performAndWait` here?  Isn't it better to not block the calling thread?

darinkrauss replied:
The `RemoteDataServicesManager` implementation requires it (as it is currently designed). It could be changed by adding a DispatchGroup to the `RemoteDataServicesManager` function, but to be consistent it would require changing all of the stores and all of the code in `RemoteDataServicesManager`.

rickpasetto commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -292,46 +304,41 @@ extension AlertStore {
 extension AlertStore {
 
-    public struct QueryAnchor<Filter: QueryFilter>: RawRepresentable, Equatable {
+    public struct QueryAnchor: RawRepresentable, Equatable {
```
Why did you get rid of the `filter` in the Anchor?  Isn't it a fundamental part of the query?

darinkrauss replied:
No, not really. 

1) The call to `executeQuery(since date: Date, ...` is only called for the first page, so the filter isn't needed agin.
2) The call to `executeQuery(fromQueryAnchor queryAnchor: QueryAnchor?, ...` doesn't need it.

The primary gotcha is that the query anchor from the second call is serialized and persisted, yet the filter was not part of that. Seemed like a big headache to serialize the filter appropriate so the code was consistent when the filter isn't used past the first call.

rickpasetto reviewed (commented):
Got some questions below, please LMK if you'd like to chat live...

Commit:
Updates based upon PR feedback

- Rename alert parameters to metadata
- Use dictionary instead of string for alert metadata

darinkrauss commented:
Updated to use dictionary rather that string to capture metadata (was parameters).

rickpasetto reviewed (approved):
👍  Thanks for doing this.  Personally I like this better... :)

Commit:
Updates based upon PR feedback

ps2 reviewed (approved):
LGTM!
